GhostNetwork.Gateway.Api/Chats/ChatsController.cs

```diff
@@ -0,0 +1,149 @@
+
+[Route("[controller]")]
+[ApiController]
+[Authorize]
+public class ChatsController : ControllerBase
+{
+    private readonly IChatStorage chatStorage;
+    private readonly ICurrentUserProvider currentUserProvider;
+
+    public ChatsController(IChatStorage chatStorage, ICurrentUserProvider currentUserProvider)
+    {
+        this.chatStorage = chatStorage;
+        this.currentUserProvider = currentUserProvider;
+    }
+
+    /// <summary>
+    /// Search user chats.
+    /// </summary>
+    /// <param name="cursor">Skip chats up to a specified id.</param>
+    /// <param name="take">Take chats up to a specified position.</param>
+    /// <returns>Filtered sequence of chats.</returns>
+    [HttpGet]
+    [ProducesResponseType(StatusCodes.Status200OK)]
+    [SwaggerResponseHeader(StatusCodes.Status200OK, Consts.Headers.Cursor, "string", "")]
+    public async Task<ActionResult<IReadOnlyCollection<Chat>>> SearchAsync(
+        [FromQuery, Range(0, int.MaxValue)] string cursor,
+        [FromQuery, Range(1, 50)] int take = 20)
+    {
+        var (chats, nextCursor) = await chatStorage.GetAsync(currentUserProvider.UserId, cursor, take);
+
+        Response.Headers.Add(Consts.Headers.Cursor, nextCursor);
+
+        return Ok(chats);
+    }
+
+    /// <summary>
+    /// Create chat.
+    /// </summary>
+    /// <param name="model">Chat model.</param>
+    /// <returns>New chat.</returns>
+    [HttpPost]
+    [ProducesResponseType(StatusCodes.Status201Created)]
+    public async Task<ActionResult<Chat>> CreateAsync([FromBody] CreateChat model)
+    {
+        model.Participants.Add(Guid.Parse(currentUserProvider.UserId));
+
+        var chat = await chatStorage.CreateAsync(model.Name, model.Participants);
+
+        return Created(string.Empty, chat);
+    }
+
+    /// <summary>
+    /// Get chat by id.
+    /// </summary>
+    /// <param name="id">Chat id.</param>
+    /// <returns>Chat.</returns>
+    [HttpGet("{id}")]
+    [ProducesResponseType(StatusCodes.Status404NotFound)]
+    [ProducesResponseType(StatusCodes.Status200OK)]
+    public async Task<ActionResult<Chat>> GetByIdAsync([FromRoute] string id)
+    {
+        var chat = await chatStorage.GetByIdAsync(id);
+
+        if (chat is null)
+        {
+            return NotFound();
+        }
+
+        if (chat.Participants.All(x => x.Id.ToString() != currentUserProvider.UserId))
+        {
+            return Forbid();
+        }
+
+        return Ok(chat);
+    }
+
+    /// <summary>
+    /// Update chat.
+    /// </summary>
+    /// <param name="id">Chat id.</param>
+    /// <param name="model">Update chat model.</param>
+    [HttpPut("{id}")]
+    [ProducesResponseType(StatusCodes.Status404NotFound)]
+    [ProducesResponseType(StatusCodes.Status204NoContent)]
+    [ProducesResponseType(StatusCodes.Status403Forbidden)]
+    [ProducesResponseType(StatusCodes.Status400BadRequest)]
+    public async Task<ActionResult> UpdateAsync([FromRoute] string id, [FromBody] UpdateChat model)
```

Comment:
Add validation (not more than 10 participants) and Chat name length from 5 to 256 chars